src/engine/symbolic_execution.rs

```diff
@@ -621,135 +789,172 @@ where
-        if bug.is_some() {
-            return Ok(bug);
-        }
+        self.check_for_valid_memory_range("write", buffer, size)?;
 
         let size_of_u64 = size_of::<u64>() as u64;
         let start = buffer / size_of_u64;
         let bytes_to_read = size + buffer % size_of_u64;
         let words_to_read = (bytes_to_read + size_of_u64 - 1) / size_of_u64;
 
         for word_count in 0..words_to_read {
-            if self.memory[(start + word_count) as usize] == Value::Uninitialized {
+            let address = start + word_count;
+
+            if self.memory[address as usize] == Value::Uninitialized {
                 trace!(
                     "write: access to uninitialized memory at {:#x} => computing reachability",
                     (start + word_count) * size_of_u64,
                 );
 
-                return self.execute_query(Query::Reachable, |info| {
-                    Bug::AccessToUnitializedMemory {
-                        info,
-                        instruction,
-                        operands: vec![],
-                    }
+                return self.execute_query(Query::Reachable).and_then(|r| {
+                    self.exit_anyway_with_bug(r, |i| {
+                        self.access_to_uninitialized_memory_bug(
+                            i,
+                            UninitializedMemoryAccessReason::ReadUnintializedMemoryAddress {
+                                description: format!(
+                                    "access to uninitialized memory at address: {:#x}",
+                                    address
+                                ),
+                                address,
+                            },
+                        )
+                    })
                 });
             }
         }
 
         self.regs[Register::A0 as usize] = Value::Concrete(size);
 
-        Ok(None)
+        Ok(())
     }
 
-    fn execute_beq_branches(
-        &mut self,
-        true_branch: u64,
-        false_branch: u64,
-        lhs: SymbolicValue,
-        rhs: SymbolicValue,
-    ) -> Result<Option<Bug>, SymbolicExecutionError> {
+    fn should_execute_branch(&mut self) -> Result<(bool, &'static str), SymbolicExecutionError> {
+        Ok(match self.execute_query(Query::Reachable)? {
+            QueryResult::Sat(_) => (true, "reachable"),
+            QueryResult::Unknown if self.options.optimistically_prune_search_space => {
+                (true, "reachability unknown")
+            }
+            _ => (false, "unreachable"),
+        })
+    }
+
+    fn with_snapshot<F, R>(&mut self, f: F) -> R
+    where
+        F: FnOnce(&mut Self) -> R,
+    {
         let memory_snapshot = self.memory.clone();
         let regs_snapshot = self.regs;
         let graph_snapshot = Box::new((*self.symbolic_state).clone());
         let brk_snapshot = self.program_break;
         let execution_depth_snapshot = self.execution_depth;
```

Comment:
Should we also save and restore `amount_of_file_descriptors` here?

Reply:
yes you're right. that would be a bug otherwise

Reply:
done